pkg/provider/aws/resources/eks.go

```diff
@@ -65,43 +111,89 @@ type (
 //
 // The method will also create a fargate profile in the default namespace and a single NodeGroup.
 // The method will create all of the corresponding IAM Roles necessary and attach all the execution units references to the following objects
-func CreateEksCluster(appName string, clusterName string, subnets []*Subnet, securityGroups []*any, units []*core.ExecutionUnit, dag *core.ResourceGraph) {
+func CreateEksCluster(cfg *config.Application, clusterName string, subnets []*Subnet, securityGroups []*any, units []*core.ExecutionUnit, dag *core.ResourceGraph) {
 	references := []core.AnnotationKey{}
 	for _, u := range units {
 		references = append(references, u.Provenance())
 	}
 
+	type groupKey struct {
+		InstanceType string
+		NetworkType  string
+	}
+
+	type groupSpec struct {
+		DiskSizeGiB int
+		refs        []core.AnnotationKey
+	}
+
+	groupSpecs := make(map[groupKey]*groupSpec)
+
+	for _, unit := range units {
+		unitCfg := cfg.GetExecutionUnit(unit.ID)
+		params := unitCfg.GetExecutionUnitParamsAsKubernetes()
+		key := groupKey{InstanceType: params.InstanceType, NetworkType: unitCfg.NetworkPlacement}
+		spec := groupSpecs[key]
+		if spec == nil {
+			spec = &groupSpec{}
+			groupSpecs[key] = spec
+		}
```

Comment:
if someone doesnt set the params for instance type then do we generate a blank spec for it?

Similarly if a unit has a requirement on disk size, but not instance type will it be ignored and not factored into the existing groups?

Reply:
may be helpful to break out the determination of node groups given a set of exec units so we can unit test it

Reply:
At this point, after defaults have been applied, the instance type and disk size shouldn't be empty / zero.

Reply:
disk size isnt in defaults it looks like but also this will cause us to create unnecessary node groups which could be wildly expensive.

Reply:
Is `aws/config.go` not the right place to put the defaults?
```go
	eksDefaults = config.KubernetesTypeParams{
		NodeType:     "fargate",
		InstanceType: "t3.medium",
		DiskSizeGiB:  20,
		Replicas:     2,
	}
```

I took these from what the default node group was in the deploylib:
```ts
    private determineNodeGroupSpecs(execUnits: EksExecUnit[]): Map<string, NodeGroupSpecs> {
        let nodeGroupSpecs: Map<string, NodeGroupSpecs> = new Map<string, NodeGroupSpecs>()
        const defaultDiskSize = 20
        const defaultInstanceType = 't3.medium'
```

> this will cause us to create unnecessary node groups

Not sure I understand why, can you go into details on that?

> see my comment about why we didnt have defaults for more info

Which comment?